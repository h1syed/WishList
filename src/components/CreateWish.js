import React, {Component} from "react";
import { API, graphqlOperation,Auth } from "aws-amplify";
import { createWish } from "../graphql/mutations";

class CreateWish extends Component {
  

       state = {
        wishOwnerId: "",
        wishOwnerUsername: "",
        wishTitle: "",
        wishBody: ""
    };
   
    

    componentDidMount = async () => {
        await Auth.currentUserInfo()
        .then(user => {
            this.setState({
                  wishOwnerId: user.attributes.sub,
                  wishOwnerUsername: user.username
            });
        });
    }

    handleChangeWish = event => this.setState({
        [event.target.name] : event.target.value 
        });

    handleAddWish = async event => {
        event.preventDefault();
        const input = {
             wishOwnerId: this.state.wishOwnerId,
             wishOwnerUsername: this.state.wishOwnerUsername,
             wishTitle: this.state.wishTitle,
             wishBody: this.state.wishBody,
             createdAt: new Date().toISOString()
        }

        await API.graphql(graphqlOperation(createWish, { input }));

        this.setState({ wishTitle: "", wishBody: ""})


   }

    render() { 
        return(
            <form className="add-wish" onSubmit={this.handleAddWish}>
                <input style={{ font: '19px'}} 
                  type="text" 
                  placeholder="New Wish"
                  name="wishTitle"
                  required
                  value={this.state.wishTitle}
                  onChange={this.handleChangeWish}
                  />

                <textarea 
                  type="text"
                  name="wishBody"
                  rows="3"
                  cols="40"
                  required
                  placeholder="Description"
                  value={this.state.wishBody}
                  onChange={this.handleChangeWish}
                  />

                <input  type="submit"
                  className="btn"
                  style={{ fontSize: '19px'}}/>
            </form>
        )
    }
}

export default CreateWish;