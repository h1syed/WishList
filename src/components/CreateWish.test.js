import React from "react";
import { shallow } from "enzyme";
import CreateWish from "./CreateWish";
import { API, Auth } from "aws-amplify";


describe("Create wish component", () => {
    const createwish = shallow(<CreateWish/>);

    it("renders properly", () => {
        expect(createwish).toMatchSnapshot();
    });

    it("renders a wish title input text area", () => {
        expect(shallow(<CreateWish/>).find({name: "wishTitle"}).length).toEqual(1);
    });
   
    it("renders a wish body input text area", () => {
        expect(shallow(<CreateWish/>).find({name: "wishBody"}).length).toEqual(1);
    });

    it("renders a submit button", () => {
        expect(shallow(<CreateWish/>).find({className: "btn"}).length).toEqual(1);
    });
    
    describe("wish title input", () => {
        it("should respond to change event and change the state of the create wish component", () => {
            const wrapper = shallow(<CreateWish/>);
            wrapper.find({name: "wishTitle"}).simulate('change', {target:{name: "wishTitle", value: "This is my wish"} });
            expect(wrapper.state("wishTitle")).toEqual("This is my wish");
        });
    });

    describe("wish body input", () => {
        it("should respond to the change event and change state of the create wish component", ()=> {
            const wrapper = shallow(<CreateWish/>);
            wrapper.find({name: "wishBody"}).simulate('change', {target: {name: "wishBody", value: "This is the wish body"}});
            expect(wrapper.state("wishBody")).toEqual("This is the wish body");
        });
    });

    describe("form" , () => {
        it("should call handleAddWish on submit", async () => {
          const wrapper = shallow(<CreateWish/>);
            const mockPreventDefault = jest.fn();
            const spy = jest.spyOn(wrapper.instance(), "handleAddWish");
            const spy1 = jest.spyOn(API, "graphql").mockReturnValue(Promise.resolve());
            wrapper.instance().forceUpdate();
            const mockEvent = {
                preventDefault: mockPreventDefault
            }
            wrapper.find({ className: "add-wish"}).simulate("submit", mockEvent);
            expect(spy).toHaveBeenCalledWith(mockEvent);
            spy.mockRestore();
        });
    });

    it("componentDidMount", async () => {
        const wrapper = shallow(<CreateWish/>);
        const instance = wrapper.instance();
        const user = {
            attributes:  {
                sub: "123"
            },
            username: "Bob"
        };
        const spy = jest.spyOn(Auth, "currentUserInfo").mockReturnValue(Promise.resolve(user));
        await instance.componentDidMount();
        expect(instance.state).toEqual({
            wishBody: "",
            wishOwnerId: "123",
            wishOwnerUsername: "Bob",
            wishTitle: ""
        });
        spy.mockRestore();
      });

   
});