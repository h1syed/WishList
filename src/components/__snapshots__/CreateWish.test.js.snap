// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create wish component renders properly 1`] = `ShallowWrapper {}`;
